src/lib.rs

```diff
@@ -506,8 +494,59 @@ impl<T, U> These<T, U> {
     /// [`These`]: enum.These.html#variant.These
     /// [`Some`]: enum.Option.html#variant.Some
     /// [`None`]: enum.Option.html#variant.Some
-    pub fn here(self) -> Option<T>
-    {
+    pub fn here(self) -> Option<T> {
+        match self {
+            These::This(t) | These::These(t, _) => Some(t),
+            These::That(_) => None,
+        }
+    }
+
+    /// Produce a [`Some<&T>`] if a [`This`] or [`These`] is found,
+    /// otherwise [`None`].
+    ///
+    /// It borrow the `These`.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use these::These;
+    ///
+    /// let these: These<&str, i8> = These::These("Hello", 42);
+    /// assert_eq!(these.here_as_deref(), Some(&"Hello"));
+    /// ```
+    ///
+    /// [`This`]: enum.These.html#variant.This
+    /// [`These`]: enum.These.html#variant.These
+    /// [`Some<&T>`]: enum.Option.html#variant.Some
+    /// [`None`]: enum.Option.html#variant.Some
+    pub fn here_as_deref(&self) -> Option<&T> {
```

Comment:
Would this be easier achieved by calling `x.here().as_ref()`? Maybe to keep with the convention of  `Option` and `Result` we should have `pub fn as_ref` as the general case, then this could be achieved by `x.as_ref().here()`. What do you think?

Reply:
The problem is that `.here()` takes complete ownership of the `These`, so whatever comes after is a little tricky.

I love the idea of implementing `.as_ref()` and `.as_mut_ref()` let me see if I can implement them.